envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,155 @@
+//     kind: ServiceRoleBinding
+//     metadata:
+//       name: test-binding-products
+//       namespace: default
+//     spec:
+//       subjects:
+//       - user: cluster.local/ns/default/sa/bookinfo-reviews
+//       - attributes:
+//           source.service: "reviews.default.svc.cluster.local"
+//           source.labels["version"]: "v1"
+//       roleRef:
+//         name: "products-viewer"
+
+// ServiceRole specification contains a list of access rules (permissions).
+// This represents the "Spec" part of the ServiceRole object. The name and namespace
+// of the ServiceRole is specified in "metadata" section of the ServiceRole object.
+message ServiceRole {
+  // Required. The set of access rules (permissions) that the role has.
+  repeated AccessRule rules = 1 [(validate.rules).repeated .min_items = 1];
+}
+
+// AccessRule defines a permission to access a list of services.
+message AccessRule {
+  // Required. A list of service names.
+  // Exact match, prefix match, and suffix match are supported for service names.
+  // For example, the service name "bookstore.mtv.svc.cluster.local" matches
+  // "bookstore.mtv.svc.cluster.local" (exact match), or "bookstore*" (prefix match),
+  // or "*.mtv.svc.cluster.local" (suffix match).
+  // If set to ["*"], it refers to all services in the namespace.
+  repeated string services = 1 [(validate.rules).repeated .min_items = 1];
+
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // Exact match, prefix match, and suffix match are supported for paths.
+  // For example, the path "/books/review" matches
+  // "/books/review" (exact match), or "/books/*" (prefix match),
+  // or "*/review" (suffix match).
+  // If not specified, it applies to any path.
+  repeated string paths = 2;
+
+  // Optional. A list of HTTP verbs (e.g., "GET", "POST").
+  // It is ignored in gRPC case because the value is always "POST".
+  // If set to ["*"] or not specified, it applies to any verb.
+  repeated string verbs = 3;
+
+  // Definition of a custom constraint. The key in the constraint refers to an attribute defined in
+  // AttributeContext (//envoy/service/auth/v2/attribute_context.proto). For example,
+  // destination.labels["app"], request.http.headers["version"].
+  message Constraint {
+    // Key of the constraint.
+    string key = 1;
+
+    // List of valid values for the constraint.
+    // Exact match, prefix match, and suffix match are supported for constraint values.
+    // For example, the value "v1alpha2" matches
+    // "v1alpha2" (exact match), or "v1*" (prefix match),
+    // or "*alpha2" (suffix match).
+    repeated string values = 2 [(validate.rules).repeated .min_items = 1];
+  }
+
+  // Optional. Extra constraints in the ServiceRole specification.
+  // The above ServiceRole examples shows an example of constraint "version".
+  repeated Constraint constraints = 4;
```

Comment:
I feel pretty strongly that we should allow both whitelist and blacklist modes, as its inevitably going to be asked for. Can we do something similar to the auth proto that we deleted?

Reply:
We need to decide how to pattern matching the attributes in an efficient way. Otherwise, the performance can be terrible. IMO, Envoy needs to process 10k requests per second, so 100us per request. We must have a performance goal when designing an infrastructure.